File: lintcode/binary_tree_paths.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    # @param {TreeNode} root the root of the binary tree
    # @return {List[str]} all root-to-leaf paths
    def binaryTreePaths(self, root):
        def helper(root, path, result):
            if not root.left and not root.right:
                result.append("->".join(path + [str(root.val)]))
            else:
                if root.left: helper(root.left, path + [str(root.val)], result)
                if root.right: helper(root.right, path+[str(root.val)], result)
            return result
        return helper(root, [], []) if root else []

File: leetcode/binary_tree_paths.js
/* Given a binary tree, return all root-to-leaf paths.
 * Time Complexity: O(n + output)
 * Space Complexity: O(h + output)
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    function helper(root, path, result) {
        if (!root.left && !root.right)
            result.push(path.concat(root.val.toString()).join("->"))
        else {
            if (root.left)
                helper(root.left, path.concat(root.val.toString()), result)
            if (root.right)
                helper(root.right, path.concat(root.val.toString()), result)
        }
        return result
    }
    return root ? helper(root, [], []) : []
};
